bokeh/models/ranges.py

```diff
@@ -119,6 +142,16 @@ class FactorRange(Range):
     this categorical range.
     """)
 
+    bound_lower = Either(String, Int, help="""
+    The minimum value that the range is allowed to go to - typically used to prevent
+    the user from panning/zooming/etc away from the data.
+    """)
+
+    bound_upper = Either(String, Int, help="""
+    The max value that the range is allowed to go to - typically used to prevent
+    the user from panning/zooming/etc away from the data.
+    """)
```

Comment:
Thanks for all this...some thoughts are brewing.

I wanted a consistent spelling across the ranges, and bokeh already has a notion of pre/suf-fixed things like `hover_line_alpha`,  `background_fill_alpha` etc.

What about a property with the suffix `_bound` or `_limit` on each of the things that end up turning into start end. So concretely:

 * Range1d
   * start - Either(Float, Datetime, Int)
   * start_bound - Either(Float, Datetime, Int, Auto, None, default=??)
   * end - Either(Float, Datetime, Int)
   * end_bound - Either(Float, Datetime, Int, Auto, None, default=??)
 * DataRange1d
  * start - Float
  * start_bound - Either(Float, Auto, None, default=Auto)
  * end - Float
  * end_bound - Either(Float, Auto, None, default=Auto)
 * FactorRange
  * factors = Either(List(String), List(Int))
  * factors_bound = Either(List(String), List(Int), Auto, None, default=Auto)

* Auto will just compute it to what your initial data is
* None will be no bound

This only leaves one question: If `factors=['bat', 'cat', 'dat', 'eat', 'fat', 'gat']` and your `factors_bound=['fat', 'cat', 'eat']`  - what is expected behavior - I am not sure. 

Ordered-like behavior: Take factors as the canonical order, then sort factors_bound into the same order and then take the uppermost and lower most elements. (As @pzwang said, factors end up in drawn-space eventually so this was inspired by that). That would leave a plot that shows `factors=['cat', 'dat', 'eat', 'fat']` and you can't pan outside that

xlim like behavior: Throw out the rows that aren't in the bound. That would leave a plot that shows `factors=['cat', 'eat', 'fat']` and you can't pan outside that.

The xlim behavior seems more intuitive to me, although I'm not sure why someone would use it. and not just set factors to what they want and use Auto for the bound.